cmd/kubeadm/app/util/error.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2014 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"os"
+	"os/user"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/renstrom/dedent"
+)
+
+const (
+	DefaultErrorExitCode = 1
+	EX_USAGE             = 64
+	EX_DATAERR           = 65
+	EX_NOINPUT           = 66
+	EX_NOUSER            = 67
+	EX_NOHOST            = 68
+	EX_UNAVAILABLE       = 69
+	EX_SOFTWARE          = 70
+	EX_OSERR             = 71
+	EX_OSFILE            = 72
+	EX_CANTCREAT         = 73
+	EX_IOERR             = 74
+	EX_TEMPFAIL          = 75
+	EX_PROTOCOL          = 76
+	EX_NOPERM            = 77
+	EX_CONFIG            = 78
+)
+
+var AlphaWarningOnExit = dedent.Dedent(`
+	kubeadm: I am an alpha version, my authors welcome your feedback and bug reports
+	kubeadm: please create issue an using https://github.com/kubernetes/kubernetes/issues/new
+	kubeadm: and make sure to mention @kubernetes/sig-cluster-lifecycle. Thank you!
+`)
+
+type NotRootError struct {
```

Comment:
I looked at your open PR #34535 and I think we could move towards a consensus between the work there and some of the ideas I have in this PR. 

I am not suggesting that we do a struct per error approach on all errors.  There are going to be some errors that we want to handle a specific way and others that are more general. In my opinion, the only way to be able to handle diverse errors in different ways (different exit codes, etc.) is not to inspect the contents of the error (something like strings.Contains()), but be able to detect what type of error they are. Which is why structs that implement the error interface would work great in this case. (See http://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully)

I personally don't like that you pass around []string, I think they easily could be a []error.  Also, in your PR, there are more than one place where errors are handled and also more than one place where you program may exit.  I think it is best practice to have one place that handles errors and determines how the program should exit (errors codes, logging messages, etc). 

If we moved towards having an []error rather than []string, one could iterate over the errors and handle each accordingly.  If some only require some logging, they just log, while if others require more action, they can do so. 

As for glog, I am not personally attached to it.  Although, logs that are more machine readable are helpful, something the default log packages doesn't handle great.  

I think in the long run, getting this right is not so much for humans as it is for automating this process.  Meaningful exit codes allow things to be scripted and the returning errors codes have a lot of value in allowing said system to be designed.